platform/pulpcore/download/single.py

```diff
@@ -101,6 +102,25 @@ def register(self, event, handler):
         """
         self.router.register(event, handler)
 
+    def clone(self):
+        """
+        Clone the download.
+        The clone has cleared:
+        - event handlers
+        - validations
+        - attachment
+        - reply
+
+        Returns:
+            Download: A cloned download.
```

Comment:
Is this method intending to do the equivalent of:

```
clone = Download(self.url, self.writer)
clone.context = deepcopy(self.context)
return clone
```
?

I'm not suggesting that's better necessarily, but I'm trying to fully understand the intent. By doing a deepcopy, you'll end up with a copy of each attribute as well. You're dropping/resetting most of them, but leaving context as-is, which is an exact copy of the original.

Reply:
Not quite the same thing.

The clone() impl will return a copy of the *concrete* class with all of the attributes except those cleared or reset.  For example:
```
download = HttpDownload(..)
clone = download.clone()
```

The *clone* is an instance of HttpDownload with all of the certificates, proxy url etc copied. Actually, `clone()` needs to be adjusted to do a shallow copy of the *context*.

Reply:
Gotcha, that makes sense. I figured you'd want the context to be shared, not copied, so thought I'd better get clarification.

The reminder that this is supposed to work with subclasses also makes a lot of sense.